### PR TITLE
Towards refactoring the api

### PR DESCRIPTION
**Pull request by joelgrondman**
refactoring of list to dict, several functions changed due to this change
quick load_from_cache class method added that automatically searched for db entry or local file for instantiating object

### CODE REVIEW

File: wordstats/cognate_db.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
+from sys import stdout
+
+import sqlalchemy.orm
+import sqlalchemy
+from sqlalchemy import Column, Integer, String, UniqueConstraint, Boolean
+
+from .base_service import Base, SimplifiedQuery
+
+
+class UnknownCognateInfo(object):
```
Comment (mircealungu) [inline wordstats/cognate_db.py]: Uh. Is this really needed? It's not used anywhere.

File: wordstats/cognate_db.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    primary = Column(String(20), nullable =False, index = True)
+    secondary = Column(String(20), nullable=False, index=True)
+
+    UniqueConstraint(word_from, word_to, primary, secondary, method)
+
+    def __init__(self, word_from, word_to, languageFrom, languageTo, method):
+        self.word_from = word_from
+        self.word_to = word_to
+        self.primary = languageFrom
+        self.secondary = languageTo
```
Comment (mircealungu) [inline wordstats/cognate_db.py]: i think we should keep a parallel structure between the word and the language. 
thus, either: 
- word_from  / language_from
- or word_primary / language_primary

File: wordstats/cognate_db.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+# structure for reviewed cognates
+class CognateWhiteListInfo(SimplifiedQuery, Base):
+    __tablename__ = 'cognate_whitelist_info'
+    __table_args__ = {'mysql_collate': 'utf8_bin'}
+
+    id = Column(Integer, primary_key=True)
+
+    word_from = Column(String(255), nullable =False, index = True)
+    word_to = Column(String(255), nullable=False, index=True)
+    primary = Column(String(20), nullable =False, index = True)
```
Comment (mircealungu) [inline wordstats/cognate_db.py]: same observations about naming as before

File: wordstats/cognate_info.py [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+    # def apply_distance_metric(self, wordlist1, wordlist2, func):
+
+    def compute(self):
+        wordlist1 = list(load_language_from_hermit(self.primary).word_info_dict.keys())
+        wordlist2 = list(load_language_from_hermit(self.secondary).word_info_dict.keys())
+
+        self.candidates = dict()
+        for w1 in wordlist1:
+            for w2 in wordlist2:
+                if self.distance_computer.edit_distance(w1, w2):
```
Comment (mircealungu) [inline wordstats/cognate_info.py]: the naming here must be improved. 
`if edit_distance(w1, w2)` does not read like a condition. 

either `if edit_distance(w1,w2) < SOMETHING` or maybe, `if very_close_distance(w1, w2):` or something else that is better than `very_close_distance` but in the same vein.

File: wordstats/cognate_info.py [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+                    if w1 in self.candidates:
+                        self.candidates[w1].append(w2)
+                    else:
+                        self.candidates[w1] = [w2]
```
Comment (mircealungu) [inline wordstats/cognate_info.py]: there are two more `pythonic` ways of doing this: 

see: https://stackoverflow.com/a/17756005/1200070

File: wordstats/cognate_info.py [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+            BaseService.session.commit()
+        except sqlalchemy.exc.IntegrityError:
+            # value to be added is already in database
+            BaseService.session.rollback()
+
+    # also save new dicts
+    @classmethod
+    def load_from_path(cls, primary, secondary, distance_function):
+        """
+        Loads cognate information given the cognate code (e.g. ennl) and method_name
```
Comment (mircealungu) [inline wordstats/cognate_info.py]: comment about cognate code is out of date.

File: wordstats/cognate_info.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+        for k, v in self.candidates.items():
+            lines.append(k + " " + " ".join(str(x) for x in v))
+
+        save_to_file(language_code_path, '\n'.join(lines))
+
+    def save_evaluation(self):
+        """
+
+        write to disk both whitelist and blacklist
+        worth a rename of the method name
```
Comment (mircealungu) [inline wordstats/cognate_info.py]: what's this "worth a rename" comment?

File: wordstats/cognate_user_evaluation.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    print("y:   whitelist")
+    print("q:   quit")
+    print("other:   blacklist")
+
+    cognateinfo = CognateInfo.load_cached(languageFrom, languageTo, method)
+    cognateinfo.save_candidates()
+    for key, values in cognateinfo.candidates.items():
+        for value in values:
+            if (key not in cognateinfo.blacklist.keys() or value not in cognateinfo.blacklist[key]) and\
+                (key not in cognateinfo.whitelist.keys() or value not in cognateinfo.whitelist[key]):
```
Comment (mircealungu) [inline wordstats/cognate_user_evaluation.py]: can we move this complicated expression to it's own function? 
what i'd like a lot would be something like: `cognate.info.is_already_evaluated(key)`,
the fact that these two lines of code use four times cognateinfo should be a hint 
that this behavior belongs to that class.

**Review by mircealungu — CHANGES_REQUESTED** [review]

File: wordstats/edit_distance.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
+import codecs
+import configparser
+from .cognate_files_path import *
+from .edit_distance_function_factory import WordDistanceFactory
+
+
+class LanguageAwareEditDistance(WordDistanceFactory):
+    def __init__(self, primary_language, secondary_language):
```
Comment (mircealungu) [inline wordstats/edit_distance.py]: short comment on what is the meaning of this class would be good. 
also, i see no language awareness in the class. why?

File: wordstats/edit_distance_function_factory.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+class WordDistanceFactory(ABC):
+
+    def __init__(self, primary_language, secondary_language):
+        super().__init__()
+        self.rules = []
+        self.threshold = 0.5
+    # in addition, derived classes need to implement the following variables
+        self.method_name = ""
+        self.languageFrom = primary_language
+        self.languageTo = secondary_language
```
Comment (mircealungu) [inline wordstats/edit_distance_function_factory.py]: here we talk about languageFrom and languageTo. 
in the other class we had primary and secondary.
we should choose *one* convention, and use it all throughout the project.

File: wordstats/edit_distance_overlap.py [hunk 11]
```diff
@@ -0,0 +1,21 @@
+import codecs
+import configparser
+from .cognate_files_path import *
+from .edit_distance_function_factory import WordDistanceFactory
+
+class WordDistanceOverlap(WordDistanceFactory):
```
Comment (mircealungu) [inline wordstats/edit_distance_overlap.py]: comment on what exactly this is would be very welcome!

**Review by mircealungu — CHANGES_REQUESTED** [review]

File: wordstats/cognate_automatic_evaluation.py [hunk 12]
```diff
@@ -0,0 +1,16 @@
+from wordstats.cognate_info import CognateInfo
```
Comment (mircealungu) [inline wordstats/cognate_automatic_evaluation.py]: comment at the top of the file specifying what it's doing.

File: wordstats/cognate_automatic_evaluation.py [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+from wordstats.edit_distance import LanguageAwareEditDistance
+
+from python_translators.translators.glosbe_over_tor_translator import GlosbeOverTorTranslator
+
+languageFrom = "de"
+languageTo = "en"
+cognateInfo = CognateInfo.load_cached(languageFrom, languageTo, LanguageAwareEditDistance)
+print(len(cognateInfo.candidates))
+#cognateInfo.blacklist.clear()
+#cognateInfo.whitelist.clear()
```
Comment (mircealungu) [inline wordstats/cognate_automatic_evaluation.py]: remove commented out code.

File: wordstats/cognate_automatic_evaluation.py [hunk 14]
```diff
@@ -0,0 +1,16 @@
+from wordstats.cognate_info import CognateInfo
+from wordstats.edit_distance import LanguageAwareEditDistance
+
+from python_translators.translators.glosbe_over_tor_translator import GlosbeOverTorTranslator
```
Comment (mircealungu) [inline wordstats/cognate_automatic_evaluation.py]: so is this GlosbeOverTor working after all? 
or is this code not used anymore?

File: wordstats/cognate_files_path.py [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+# useful methods for handling files
+
+def path_of_cognate_languages(primary, secundary):
+    package_directory = os.path.dirname(os.path.abspath(__file__))
+    file_path = package_directory + os.sep + "{0}{1}{2}{3}".format(DATA_FOLDER_COGNATES, os.sep, primary, secundary)
+    return file_path
+
+def path_of_cognate_candidates(primary, secundary, method_name):
+
+    return path_to_cognate_file(primary, secundary, CANDIDATES, method_name)
```
Comment (mircealungu) [inline wordstats/cognate_files_path.py]: secondary not secundary.
here and hereafter.

File: wordstats/cognate_files_path.py [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+    return path_to_cognate_file(primary, secundary,
+                                WHITELIST if len(author) == 0 else (WHITELIST + "_" + author))
+
+
+def path_of_cognate_rules(primary, secundary, method_name, author:str = ""):
+
+    return path_to_cognate_file(primary, secundary, RULES if len(author) == 0 else (RULES + "_" + author)
+                                , method_name)
+
+def path_to_cognate_file(primary, secundary, file_name, method_name = ""):
```
Comment (mircealungu) [inline wordstats/cognate_files_path.py]: this method is private to this file (or module in Python). 
the convention is to prefix it's name with an underscore: `_ path_to_cognate_file`.

File: wordstats/cognate_files_path.py [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+    if method_name is not "":
+        file_path += method_name + os.sep
+    file_path += file_name + ".txt"
+
+    os.makedirs(os.path.dirname(file_path), exist_ok=True)
+
+    return file_path
+
+
+def load_from_path(path):
```
Comment (mircealungu) [inline wordstats/cognate_files_path.py]: truth is, this file could be split in two: 
- cognate_paths
- file_operations

and moved to it's own module: file_handling probably.

File: wordstats/edit_distance_translate.py [hunk 18]
```diff
@@ -0,0 +1,36 @@
+import codecs
+import configparser
+from .cognate_files_path import *
+from .edit_distance_function_factory import WordDistanceFactory
+from googletrans import Translator
```
Comment (mircealungu) [inline wordstats/edit_distance_translate.py]: what's this googletrans thing? 
it's not in the required packages in setup.py.

File: wordstats/cognate_info.py [hunk 19]
```diff
@@ -0,0 +1,391 @@
...
+
+    def best_guess(self):
+        best_dict = dict(self.candidates)
+        for key in self.blacklist.keys():
+            best_dict.pop(key)
+
+        return best_dict
+
+    # generates candidates based on distance function func and word lists
+    # def apply_distance_metric(self, wordlist1, wordlist2, func):
```
Comment (mircealungu) [inline wordstats/cognate_info.py]: what's this? 
is it a comment for a commented out function? 
remove both.

File: wordstats/cognate_info.py [hunk 20]
```diff
@@ -0,0 +1,391 @@
...
+
+        self.candidates = defaultdict(list)
+        i = 0
+        for w1 in wordlist1:
+            for w2 in wordlist2:
+                if self.distance_computer.is_candidate(w1, w2):
+                    self.candidates[w1].append(w2)
+
+
+    def compute_translator(self, translator:Translator, save:Boolean = False):
```
Comment (mircealungu) [inline wordstats/cognate_info.py]: this is not a self-explanatory name.
think of a better name, and also add docstring.

File: wordstats/edit_distance.py [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+class LanguageAwareEditDistance(WordDistanceFactory):
+    def __init__(self, primary, secondary, author:str = ""):
+        super().__init__(primary, secondary, author)
+        self.method_name = "edit_distance"
+        self.threshold = 0.5
+        self._initialize_distances()
+
+    def _initialize_distances(self):
+
+        # these might change based on the primary secondayr
```
Comment (mircealungu) [inline wordstats/edit_distance.py]: secondayr

File: wordstats/edit_distance.py [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+        self.replace_distance = 1
+        self.add_distance = 1
+
+    def edit_distance_function(self, word1: str, word2: str):
+
+        lengthLongest = max(len(word1),len(word2))
```
Comment (mircealungu) [inline wordstats/edit_distance.py]: length_longest

File: wordstats/edit_distance_translate.py [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+
+class WordDistanceOverlap(WordDistanceFactory):
```
Comment (mircealungu) [inline wordstats/edit_distance_translate.py]: `WordDistanceOverlap`, a kind of `WordDistance`, can't be the subclass of a `WordDistanceFactory`. 
This makes me think that your superclass should be actually named `WordDistance`. 
Would that make sense to you?

**Review by mircealungu — COMMENTED** [review]
see the comments.
also, i'll send you in a bit a pull request to your branch, with a few restructurings.

**Comment by joelgrondman** [comment]
I've merged the branches together, restructurings are applied as well as the renames.

**Comment by mircealungu** [comment]
waiting for a smaller PR :)

**Pull request closed without merging**